lib/resultuploaders/s3/s3.rb

```diff
@@ -0,0 +1,143 @@
+# frozen_string_literal: true
+
+module AutoHCK
+  # S3 class for managing interactions with S3-compatible storage services.
+  # This class is designed to work with services that are compatible with the S3 API,
+  # including AWS S3 and Alibaba Cloud OSS.
+  class S3
+    CONFIG_JSON = 'lib/resultuploaders/s3/s3.json'
+    INDEX_FILE_NAME = 'index.html'
+
+    include Helper
+
+    # Provides read access to the URL for the index.html file.
+    #
+    # In the context of S3 compatibility, there is no concept of folders.
+    # The `url` will point to an index.html file that records the paths of all files.
+    # This allows `index.html` to be displayed directly in HTML format.
+    #
+    # Note: While AWS S3 supports this behavior of rendering HTML files directly,
+    # not all S3-compatible storage services do. For example, Alibaba Cloud OSS
+    # has security restrictions on its default domain that may require files to be
+    # downloaded rather than rendered directly, thus preventing HTML files from being displayed.
```

Comment:
The documentation says it is possible to host a static website:
https://www.alibabacloud.com/help/en/oss/configure-static-website-hosting

Reply:
Building on your previous comment (suggesting the use of Aws::S3::Object#key instead of the URL I used), I have a clearer understanding of your point. You are suggesting that we access the bucket by configuring it as a static website, which aligns with the solution I previously referred to as option 1. However, my current implementation is actually option 2, where I simulate a static webpage using index.html rather than using a true static webpage. The advantage of this approach is that users do not need to perform any additional configuration, but the trade-off is that they can only access it through the S3 default domain.

![image](https://github.com/user-attachments/assets/8f402981-1ccb-4ffc-b183-ebc932a51297)

Regarding the comment about the access restrictions with Alibaba Cloud, I am specifically addressing the limitations related to their default domain. For files with an .html extension, Alibaba Cloud, for security reasons, does not open them in text/html format in the browser; instead, it triggers a download. You can find more details in the following links: [Alibaba Cloud Documentation 1](https://www.alibabacloud.com/help/en/oss/user-guide/overview-71?spm=a2c63.p38356.0.0.6d6836b3n6E2YQ) and [Alibaba Cloud Documentation 2](https://www.alibabacloud.com/help/en/oss/user-guide/how-to-ensure-an-object-is-previewed-when-you-access-the-object?spm=a2c63.p38356.0.0.2a5512b5Bu7J3v). However, if we aim for full support for Alibaba Cloud, it may indeed be more appropriate to choose option 1, allowing users to configure a custom domain and thereby bypass this restriction. Of course, if we switch to option 1, some modifications to my current code will be necessary. If my understanding is incorrect, please let me know.

Reply:
However, I must say that using option 1 can indeed be somewhat cumbersome, as it requires users to register a custom domain, which involves additional costs and processes like record-filing. If users do not opt for a custom domain, they will still face the same issues associated with the default domain used in option 2. If one can accept that accessing Alibaba Cloud OSS URLs via the browser results in downloads instead of displaying them in HTML format, option 2 also seems like a viable choice. What do you think? @akihikodaki @kostyanf14

Reply:
@zjmletang @akihikodaki 
I agree with option 2 to have index.html. Hosting a static website is more complicated and requires specific configuration. I want to leave this to a user who will use the S3 uploader. If in the future we get other requests from the community/customers we can update this uploader.

Reply:
I suggested using `Aws::S3::Object#key` for the user-visible label instead of parsing URL; it is irrelevant with this comment.
I thought the static website hosting allows using the default domain as it has examples such as `examplebucket.oss-cn-hangzhou.aliyuncs.com`, but apparently that's not true. Let's leave the comment.

Reply:
@akihikodaki , Okay, I understand your point. I have two small details I would like to confirm with you:

1.When you mentioned "key," do you mean the full path or just the last part of the file name? I feel that using only the last part of the file name might be more appropriate, as the full path can be quite long.

2.In terms of implementation, I will change the original @uploader_urls to @object_keys = [] to store the key for each object (I won't directly use the objects array because it still needs to handle duplicates based on keys, which may affect performance).

Then the implementation for index.html.erb is as follows:
![image](https://github.com/user-attachments/assets/eab484a6-76ce-46c4-b685-01c5540e152f)

Reply:
> It would be better to store just the last part of the file names.

@akihikodaki ,The reason I am not using a filename array is that if I do, I would need both urls and filenames arrays simultaneously, and in the ERB, I would need to associate the elements of the two arrays (although this association might be quite simple here). On the other hand, if I use keys or my original urls, I only need one array. What do you think?

Reply:
@zjmletang You only need one array either way; create URLs from filenames.

Reply:
@akihikodaki ，Okay, I misinterpreted your point earlier， I'm not sure if I understood your point now. Could you please confirm: do you mean to pass both @path and @filename into the ERB to construct a key, and then create an object out of that? The URL would still be obtained through public_url, just like in the image I shared above